### PR TITLE
libbuitin/compiler_rt: Supports separate builtin rt.builtins or rt.profile

### PR DESCRIPTION
**Pull request by W-M-R**
*Note: Please adhere to [Contributing Guidelines](https://github.com/apache/nuttx/blob/master/CONTRIBUTING.md).*

## Summary
1.  Organize make.defs and eliminate compilation warnings'
2.  Supports separate builtin rt.builtins or rt.profile
## Impact
None
## Testing
1. enable CONFIG_COMPILER_RT_BUILTINS to builtin rt.builtins instead of the toolchain's own
2. enable CONFIG_COMPILER_RT_PROFILE to builtin rt.profile instead of the toolchain's own

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No, this PR does not fully meet the NuttX requirements.  While it provides a basic summary, it lacks crucial details.

Here's a breakdown of what's missing:

* **Insufficient Summary:**  "Organize make.defs and eliminate compilation warnings" is too vague. *Which* warnings? *How* are they being eliminated?  Similarly, "Supports separate builtin rt.builtins or rt.profile" needs more explanation. What does "separate builtin" mean?  Why is this change necessary? What problem does it solve? What was the behavior before?
* **Missing Impact Assessment:**  Stating "None" for all impact categories is unlikely.  Even seemingly minor changes can have unforeseen consequences.  Consider the impact on:
    * **Users:** Will they need to change any configuration options?  Will the behavior of any existing functionality change?
    * **Build:** Does the build process take longer or shorter?  Are there new dependencies?
    * **Hardware:**  Even if the code targets a specific architecture, the changes might affect others indirectly.  Explicitly state which architectures were tested and confirm others are unaffected.
    * **Documentation:** If the build system is changing, the documentation likely needs updates.
* **Missing Testing Details:** The testing section lacks crucial information:
    * **Build Host(s):**  Operating system, compiler version, etc. are missing.
    * **Target(s):** Be specific about the architecture and board used for testing.  "sim, RISC-V, ARM" is too broad.  Provide concrete examples like "qemu-rv32ima", "STM32F4 Discovery".
    * **Logs:** The provided "testing logs" sections are empty.  Include relevant logs demonstrating the issue before the change and the improved behavior after the change.  Show that the warnings are indeed eliminated.


**Example of a better Summary:**

> This PR addresses compiler warnings arising from inconsistencies in `make.defs` when building with certain toolchains (e.g., GCC 12.x). It reorganizes the definition of specific macros (list the macros) to ensure consistent usage and eliminates warnings related to (describe the warnings, e.g., unused variables, implicit conversions).  Additionally, it introduces `CONFIG_COMPILER_RT_BUILTINS` and `CONFIG_COMPILER_RT_PROFILE` options.  These allow users to selectively link against NuttX's built-in implementations of `rt.builtins` and `rt.profile` rather than relying on the toolchain's versions. This improves compatibility with certain toolchains where the built-in versions are missing or cause conflicts.



By providing more specific details and addressing each point in the requirements, the PR will be much easier to review and ultimately merge.

File: libs/libbuiltin/Kconfig [hunk 1]
```diff
@@ -29,12 +29,20 @@ config COMPILER_RT_VERSION
 	string "Select LLVM Compiler-rt version"
 	default "17.0.1"
 
+config COMPILER_RT_BUILTINS
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:32]: why not reuse BUILTIN_COMPILER_RT

**Comment by W-M-R** [comment]
Since the previous modification of compiler_rt/cmake was omitted, I will complete it in the next patch

File: libs/libbuiltin/Kconfig [hunk 2]
```diff
@@ -5,36 +5,42 @@
...
 
+config LIB_BUILTIN
+	bool "Builtin the toolchain library"
+
+config LIB_COMPILER_RT
+	bool "Enable LLVM Compiler-rt library support"
+
 choice
-	prompt "Builtin toolchain support"
+	prompt "Builtin toolchain libgcc"
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:15]: revert the change

File: libs/libbuiltin/Kconfig [hunk 3]
```diff
@@ -5,36 +5,42 @@
 
 comment "Toolchain Library Support"
 
+config LIB_BUILTIN
+	bool "Builtin the toolchain library"
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:9]: remove the prompt string

File: libs/libbuiltin/Kconfig [hunk 4]
```diff
@@ -5,36 +5,42 @@
...
+
+config LIB_COMPILER_RT
+	bool "Enable LLVM Compiler-rt library support"
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:12]: remove too

File: libs/libbuiltin/Kconfig [hunk 5]
```diff
@@ -5,36 +5,42 @@
...
+
 choice
-	prompt "Builtin toolchain support"
+	prompt "Builtin toolchain libgcc"
 	default BUILTIN_TOOLCHAIN
 	---help---
-		Choose to compile libraries related to toolchain into the OS
+		Select the libgcc that needs to be builtin
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:18]:
```suggestion
		Select the builtin library
```

File: libs/libbuiltin/Kconfig [hunk 6]
```diff
@@ -5,36 +5,42 @@
...
+	select LIB_COMPILER_RT
 	---help---
-		Compile the LLVM Compiler-rt library into the OS.
+		Compile the LLVM rt.builtins library into the OS.
 
 config BUILTIN_TOOLCHAIN
-	bool "Link the toolchain library"
-	---help---
-		Use the toolchain library that comes with the compiler
+	bool "Link toolchain libgcc library to the OS"
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:28]:
```suggestion
	bool "Link toolchain builtin library to the OS"
```

File: libs/libbuiltin/Kconfig [hunk 7]
```diff
@@ -5,36 +5,42 @@
...
-config COMPILER_RT_VERSION
+config LIB_COMPILER_RT_VERSION
 	string "Select LLVM Compiler-rt version"
+	depends on LIB_COMPILER_RT
 	default "17.0.1"
 
-config COMPILER_RT_HAS_BFLOAT16
+config LIB_COMPILER_RT_HAS_BFLOAT16
 	bool "Enable support for bfloat16 in Compiler-rt"
+	depends on BUILTIN_COMPILER_RT
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:45]:
```suggestion
	depends on LIB_COMPILER_RT
```

File: libs/libbuiltin/Kconfig [hunk 8]
```diff
@@ -5,36 +5,42 @@
...
 config BUILTIN_TOOLCHAIN
-	bool "Link the toolchain library"
-	---help---
-		Use the toolchain library that comes with the compiler
+	bool "Link toolchain libgcc library to the OS"
 
 endchoice
 
-if BUILTIN_COMPILER_RT
+config BUILTIN_COMPILER_RT_PROFILE
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:32]:
```suggestion
config COVERAGE_COMPILER_RT
```

**Review by xiaoxiang781216 — APPROVED** [review]
